include/engine/routing_algorithms/one_to_many.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef OSRM_ONE_TO_MANY_HPP
+#define OSRM_ONE_TO_MANY_HPP
+
+#include "engine/routing_algorithms/routing_base.hpp"
+#include "engine/search_engine_data.hpp"
+#include "util/typedefs.hpp"
+#include "util/simple_logger.hpp"
+
+#include <boost/assert.hpp>
+
+#include <limits>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+namespace osrm
+{
+namespace engine
+{
+namespace routing_algorithms
+{
+
+template <class DataFacadeT>
+class OneToManyRouting final
+    : public BasicRoutingInterface<DataFacadeT, OneToManyRouting<DataFacadeT>>
+{
+    using super = BasicRoutingInterface<DataFacadeT, OneToManyRouting<DataFacadeT>>;
+    using QueryHeap = SearchEngineData::QueryHeap;
+    using PredecessorMap = std::unordered_map<NodeID, NodeID>;
+    using DistanceMap = std::unordered_map<NodeID, EdgeWeight>;
+    SearchEngineData &engine_working_data;
+
+  public:
+    OneToManyRouting(DataFacadeT *facade, SearchEngineData &engine_working_data)
+        : super(facade), engine_working_data(engine_working_data)
+    {
+    }
+
+    ~OneToManyRouting() {}
+
+    void operator()(PhantomNode &phantomSource,
+                    const int distance,
+                    PredecessorMap &predecessorMap,
+                    DistanceMap &distanceMap) const
+    {
+
+        engine_working_data.InitializeOrClearFirstThreadLocalStorage(
+            super::facade->GetNumberOfNodes());
+
+        QueryHeap &query_heap = *(engine_working_data.forward_heap_1);
+        QueryHeap &backward_heap = *(engine_working_data.reverse_heap_1);
+
+        query_heap.Clear();
+        // insert target(s) at distance 0
+        if (SPECIAL_NODEID != phantomSource.forward_node_id)
+        {
+            query_heap.Insert(phantomSource.forward_node_id,
+                              -phantomSource.GetForwardWeightPlusOffset(),
+                              phantomSource.forward_node_id);
+        }
+        if (SPECIAL_NODEID != phantomSource.reverse_node_id)
+        {
+            query_heap.Insert(phantomSource.reverse_node_id,
+                              -phantomSource.GetReverseWeightPlusOffset(),
+                              phantomSource.reverse_node_id);
+        }
+
+        util::SimpleLogger().Write() << phantomSource;
+        // explore search space
+        while (!query_heap.Empty())
+        {
+//            ForwardRoutingStep(query_heap, distance, predecessorMap, distanceMap);
+            ForwardRoutingStep(query_heap, backward_heap, )
+        }
+    }
+
+    void ForwardRoutingStep(QueryHeap &query_heap,
```

Comment:
If I understand your code correctly, it will not in-fact compute an actual isochrone.
Since you are operating on a graph hierarchy, the OSRM graph will only contain upwards facing edges.
If you start your search on the topmost node, your search will return an empty set of edges.
The rest will depend severely on the hierarchy itself.

I recommend reading up on Contraction Hierarchies (CH) in general. In short: to find actual distances in a CH, you require a bidirectional search. Variants of One-To-Many exist, but they require the addition of downwards-facing graphs (which are not available in OSRM).
The output of this search is more or less random.